### PR TITLE
Part 12. Auction response stage

### PR DESCRIPTION
**Pull request by legendko**

### CODE REVIEW

File: exchange/bidder.go [hunk 1]
```diff
@@ -54,7 +55,7 @@ type AdaptedBidder interface {
 	//
 	// Any errors will be user-facing in the API.
 	// Error messages should help publishers understand what might account for "bad" bids.
-	requestBid(ctx context.Context, bidderRequest BidderRequest, conversions currency.Conversions, reqInfo *adapters.ExtraRequestInfo, adsCertSigner adscert.Signer, bidRequestOptions bidRequestOptions, alternateBidderCodes openrtb_ext.ExtAlternateBidderCodes) ([]*pbsOrtbSeatBid, []error)
+	requestBid(ctx context.Context, bidderRequest BidderRequest, conversions currency.Conversions, reqInfo *adapters.ExtraRequestInfo, adsCertSigner adscert.Signer, bidRequestOptions bidRequestOptions, alternateBidderCodes openrtb_ext.ExtAlternateBidderCodes, hookExecutor hookexecution.StageExecutor) ([]*pbsOrtbSeatBid, []error)
```
Comment (VeronikaSolovei9) [inline exchange/bidder.go]: This change is probably a merge issue, for AuctionResponseStage `hookExecutor hookexecution.StageExecutor` is not needed here (it's needed for other stages). May you please check this, looks like this PR can be smaller. Unless I miss something.

File: exchange/exchange.go [hunk 2]
```diff
@@ -397,7 +397,9 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 	}
 
 	// Build the response
-	return e.buildBidResponse(ctx, liveAdapters, adapterBids, r.BidRequestWrapper.BidRequest, adapterExtra, auc, bidResponseExt, cacheInstructions.returnCreative, r.ImpExtInfoMap, errs)
+	bidResponse, err := e.buildBidResponse(ctx, liveAdapters, adapterBids, r.BidRequestWrapper.BidRequest, adapterExtra, auc, bidResponseExt, cacheInstructions.returnCreative, r.ImpExtInfoMap, errs)
+
+	return bidResponse, err
```
Comment (VeronikaSolovei9) [inline exchange/exchange.go:410]: Good change! I've done it couple of times locally to simplify debugging, haha

File: exchange/exchange.go [hunk 3]
```diff
@@ -546,7 +549,7 @@ func (e *exchange) getAllBids(
 				addCallSignHeader:   isAdsCertEnabled(experiment, e.bidderInfo[string(bidderRequest.BidderName)]),
 				bidAdjustments:      bidAdjustments,
 			}
-			seatBids, err := e.adapterMap[bidderRequest.BidderCoreName].requestBid(ctx, bidderRequest, conversions, &reqInfo, e.adsCertSigner, bidReqOptions, alternateBidderCodes)
+			seatBids, err := e.adapterMap[bidderRequest.BidderCoreName].requestBid(ctx, bidderRequest, conversions, &reqInfo, e.adsCertSigner, bidReqOptions, alternateBidderCodes, hookExecutor)
```
Comment (VeronikaSolovei9) [inline exchange/exchange.go:560]: This change will go away too.

File: exchange/bidder.go (hunk 1 above)
Comment (legendko) [reply exchange/bidder.go]: Yes, you are correct. I guess I moved it by cherry-picking together with some other useful changes. But the hookExecutor is not needed there.

File: exchange/exchange.go (hunk 2 above)
Comment (legendko) [reply exchange/exchange.go:410]: You figured me out :D

File: exchange/exchange.go [hunk 4]
```diff
@@ -302,7 +302,7 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 			alternateBidderCodes = *r.Account.AlternateBidderCodes
 		}
 
-		adapterBids, adapterExtra, anyBidsReturned = e.getAllBids(auctionCtx, bidderRequests, bidAdjustmentFactors, conversions, accountDebugAllow, r.GlobalPrivacyControlHeader, debugLog.DebugOverride, alternateBidderCodes, requestExt.Prebid.Experiment)
+		adapterBids, adapterExtra, anyBidsReturned = e.getAllBids(auctionCtx, bidderRequests, bidAdjustmentFactors, conversions, accountDebugAllow, r.GlobalPrivacyControlHeader, debugLog.DebugOverride, alternateBidderCodes, requestExt.Prebid.Experiment, r.HookExecutor)
```
Comment (VeronikaSolovei9) [inline exchange/exchange.go:311]: Thank you for updates! This one is not needed here too. And Line 517. This should also reduce unit test changes in the "exchange" package.

Comment (4lexvav) [reply exchange/exchange.go:311]: Updated, removed all unnecessary code within this PR

**Comment by VeronikaSolovei9** [comment]
Thank you for the updates. Code looks good! Ready to approve it after merge conflicts are fixed.

**Comment by AlexBVolcy** [comment]
This also looks good to me, and will approve after merge conflicts are resolved!

**Comment by legendko** [comment]
merged

**Pull request closed without merging**